### PR TITLE
[4190] - publicly accessible monthly reporting csv downloads

### PR DESCRIPTION
**Pull request by avinhurry**
## Context

Adding publically accessible monthly reporting CSV downloads as an alternative to generating these in the Monthly report cron job (which is behind auth).

[Kick off doc](https://docs.google.com/document/d/17escaVH_B2RHH52bDH3FvwZHTQvRosH21mYjnRQJR50)

## Changes proposed in this pull request

- Add a new filename helper method to use just the month and year in the filename

- Add CSV download logic to `monthly_statistics_controller`

- Add dynamic url for the download links (based on `year`, `month` and `export_type`)

- Include `year` and `month` in filename

- Add html with all the download links above footnotes and change `applicant` to `candidate`

**Before**
image

**After**
image

## Guidance to review

- Navigate to `publications/monthly-statistics`
- Check out the CSV download links in the `Download the data` section 

**If you are unable to get on to this page, try following the steps below:**

- Comment out `return false unless MonthlyStatisticsTimetable.generate_monthly_statistics?
` in `generate_monthly_statistics.rb`

- Run Sidekiq in a separate terminal by executing `bundle exec sidekiq`

- Execute `GenerateMonthlyStatistics.new.perform` in a rails console (this needs sidekick running to run the job it spawns) 
---------------------------------------------------------------------------------------------------------------------
- ~~It is possible to 'hack' the URL to change the filename - Is this something we are ok with considering this now publically downloadable?~~
- ~~Do we require the two links in the screenshot below as they don't seem to be available using this new approach?~~

image

- ~~Should we be consistent with our terminology, i.e `candidate` vs `applicant`?~~
- ~~Does anyone know a nice efficient way to calculate and add the size of the CSV downloads to the download link?~~

### CODE REVIEW

**Comment by avinhurry** [comment]
[Monthly report: support public CSV downloads](https://trello.com/c/CbYYrQlk/4190-monthly-report-support-public-csv-downloads)

File: app/controllers/publications/monthly_statistics_controller.rb [hunk 1]
```diff
@@ -10,6 +10,12 @@ def show
       @exports = MonthlyStatisticsTimetable.current_exports
     end
 
+    def download
+      data_export = DataExport.find_by(export_type: params[:export_type])
```
Comment (duncanjbrown) [inline app/controllers/publications/monthly_statistics_controller.rb:14]: As discussed, we can pull the same data out of the `MonthlyStatisticsReport`

File: spec/system/monthly_statistics/monthly_statistics_page_spec.rb [hunk 2]
```diff
@@ -6,14 +6,74 @@
     FeatureFlag.activate('publish_monthly_statistics')
     create_application_choices
     create_monthly_stats_report
+    visit_monthly_statistics_page
   end
 
-  scenario 'User visits monthly statistics page' do
-    when_i_visit_the_monthly_statistics_page
+  scenario 'User views monthly statistics page' do
```
Comment (duncanjbrown) [inline spec/system/monthly_statistics/monthly_statistics_page_spec.rb:12]: As discussed, we can keep this spec slim and happy — perhaps we could have a request spec or something which checked all the reports worked and they came back with correct `Content-Type` etc?

**Review by duncanjbrown — COMMENTED** [review]
Discussed offline, notes inline! We could remove the DataExport generation in this PR or a subsequent one

File: app/controllers/publications/monthly_statistics_controller.rb [hunk 3]
```diff
@@ -6,10 +6,38 @@ def show
...
+      data = SafeCSV.generate(raw_data['rows'].map(&:values), header_row)
+      send_data data, filename: export_filename, disposition: :attachment
     end
 
     def redirect_unless_published
       redirect_to root_path unless FeatureFlag.active?(:publish_monthly_statistics)
     end
+
+    def valid_date?
+      params[:date] == Time.zone.today.strftime('%Y-%m')
```
Comment (duncanjbrown) [inline app/controllers/publications/monthly_statistics_controller.rb:32]: This means they can only get reports for the current calendar month, but the current report is for the previous calendar months (ie the report date now is `2021-11`)

File: app/views/publications/monthly_statistics/show.html.erb [hunk 4]
```diff
@@ -269,7 +275,23 @@
...
-    <h2 class="govuk-heading-l govuk-!-padding-top-4" id="footnotes">12. Footnotes</h2>
+    <h2 class="govuk-heading-l govuk-!-padding-top-4" id="downloads">12. Download the data</h2>
+
+    <p class="govuk-body">You can download all of the data on this page in CSV (comma separated values) format, which can be opened in a spreadsheet and used in other analysis tools.</p>
+
+    <p class="govuk-body">The downloads include some further breakdowns of the data by combining fields - for example, applicant age and sex.</p>
+
+    <% @report.each do |export_type, size| %>
+      <p class='govuk-body'>
+    <%= govuk_link_to t("#{export_type}.label") + " #{size.to_s(:human_size)}", publications_monthly_report_download_path(export_type: export_type, date: Time.zone.today.strftime('%Y-%m')) %>
```
Comment (duncanjbrown) [inline app/views/publications/monthly_statistics/show.html.erb:286]:
```suggestion
      <%= govuk_link_to t("#{export_type}.label") + " #{size.to_s(:human_size)}", publications_monthly_report_download_path(export_type: export_type, date: Time.zone.today.strftime('%Y-%m')) %>
```

Comment (duncanjbrown) [reply app/views/publications/monthly_statistics/show.html.erb:286]: Better to hardcode `Time.zone.today.strftime('%Y-%m')` to `2021-11` for now, until we have the data changes in (ie when we know which month's report we're after)?

File: config/locales/monthly_statistics.yml [hunk 5]
```diff
@@ -1,25 +1,21 @@
 en:
-  monthly_statistics_candidates_by_status:
-    label: Applicants by status (CSV)
-  monthly_statistics_applications_by_status:
+  candidates_by_status:
+    label: Candidates by status (CSV)
```
Comment (duncanjbrown) [inline config/locales/monthly_statistics.yml:3]: Do the keys now no longer match the keys in the monthly report hash?

File: app/views/publications/monthly_statistics/show.html.erb (hunk 4 above)
Comment (avinhurry) [reply app/views/publications/monthly_statistics/show.html.erb]: thanks - changes made

File: app/controllers/publications/monthly_statistics_controller.rb (hunk 3 above)
Comment (avinhurry) [reply app/controllers/publications/monthly_statistics_controller.rb]: thanks - changes made

**Review by duncanjbrown — APPROVED** [review]

**Pull request merged**